### ISSUE TITLE
General discussion of development

### ISSUE THREAD

**Issue by not-matt** [not-matt/Systematic-LEDs#3]
Please leave any thoughts or comments on the movement of the program here.

**Comment by not-matt** [comment]
Most major issues should be resolved now, including "Scroll" and "Energy" causing crashes, and the firmware problems.

To do still - That one pesky LED!

**Comment by not-matt** [comment]
Any features or ideas you would like to see implemented, either email me or drop it here 👍

**Comment by Gaerisson** [comment]
+ Static effect: Stroboscope, I actually worked on it but it looked weird, this effect can be usefull
> Auto Mode (I saw that you tried something)
~ Random color option like before for **Scroll** effect
~ API to control LED from your software
# Logo for your software :)

**Comment by not-matt** [comment]
- Stroboscope. Sounds cool, would you explain what it should look like?
- Auto mode. Yeah... maybe. It was left as unfinished because I couldn't figure out a way to make it look good. 
- Random Scroll colour. Sure thing, do you mean randomisation of the three colour options?
- API will come hand in hand with the web server. I'll try make it flexible so anyone can use it.
- Logo is underway ;)

**Comment by Gaerisson** [comment]
I'll show you later what I think with Strobe Effect, and yeah for random scroll color 😄 the 3 colors looks cool but it will be better with a checkbox random color, and another suggestion: a way to control the brightness of LEDs from the software

**Comment by Gaerisson** [comment]
![image](https://user-images.githubusercontent.com/25458928/37107110-bf48c106-2233-11e8-986c-1914b96e8ece.png)
the settings should looks like this and the effect, like this https://www.youtube.com/watch?v=8izCmFnciNw (little bit glitchy) I don't really know if this can destroy LED, I don't think so, but it can be good for party :P

**Comment by not-matt** [comment]
That's pretty much what beat does, but I could flesh it out a bit more.
Different colours for different frequencies, use a gradient rather than
single colour, that kind of thing.

**Comment by Gaerisson** [comment]
yeah but the difference between beat effect and Strobe is that strobe will not be with the music, just flickering like a stroboscope

**Comment by Gaerisson** [comment]
another little things:
+ when we close the software, do something that's turning off LED
+ minimize the software to the taskbar (really usefull)
And I think that's all for me, for the moment ^^ I told you all my ideas 😀

**Comment by kennerblick** [comment]
I would like to run the tool on a Raspberry or Odroid without GUI as a service. Is there a way to pass the light mode as an argument?

**Comment by kennerblick** [comment]
Could you extend it to the function of controlling Philips Hue lights?

**Comment by not-matt** [comment]
Yep that's definitely possible but you'll have to make a bit yourself.

First, in config.py, set "use_gui" to false. Then to change effect, you
need to write something to modify the variable stored in
config.settings["devices"][(board
name)]["configuration"]["current_effect"].

Available effects can be found in Visualizer.effects.

Let me know if you have any luck. I'll be adding similar functionality at
some point, probably looking at an API which will let you do this much more
easily.

Matthew

**Comment by not-matt** [comment]
Open a new issue, I'll have to get back to you on that one!

**Comment by luxxl** [comment]
Fade and gradient rolling speeds needs to be fixed in my opinion it's too fast. Would be nice to make that slower and/or add more steps to slider so you could get full cycle like in minute/half minute/couple seconfs etc.

**Comment by ahodges9** [comment]
We should figure out how we want the controller configuration settings (and future web interface) to interact with the client devices. With the support for the ESPixelStick FW we can actually remove some duplication of configuration values. Things like the number of leds, port configuration, etc can all be stored on the clients and they can just send their capabilities when connecting to the controller. So you don't have to worry about keeping things like LED counts in sync instead these are properties of the client that the controller can query via HTTP/GET or WebSockets.

We can also have auto-detection where no configuration at all is need to be stored on the controller (not even IP). The controller just issues a HTTP/GET on all devices and those that respond with the correct configuration properties are automatically pulled in.

**Comment by charlieali** [comment]
What about assigning different colors to different frequency ranges? Like lets say you make anything 200mhz -1000mhz blue 1001mhz-1500mhz green or something like that. Scott had implemented a cycling in the experimental code on the other branch that was cool where it cycled through colors.  Maybe beat detection to cycle through the colors?

**Comment by not-matt** [comment]
Interesting idea! I have been thinking about something similar, a little
like "energy" but the whole strip. Cheers :)

**Comment by MattBrinza** [comment]
I've actually been working on this silently.  For my own version, I added keyboard controls to toggle between modes with a mapped controller, and I have implemented a feature that randomizes the colors/pattern when songs change (need to download snip for this to work).  Basically, its like a "no screen DJ" mode.   Let me know if anyone is interested in looking at it.

**Comment by not-matt** [comment]
Sounds awesome! What controller do you use?

**Comment by fht4gn** [comment]
@MattBrinza  mapping to a midi controller? That is amazing feature!

**Comment by MattBrinza** [comment]
I should clarify.  I mapped a lot of the functions (on / off, solid color modes, scroll through visualizer modes) to keyboard inputs, then I use a different program to map an xbox controller to those keyboard inputs.  I have them all in the config file so people can customize what they want the key bindings to be.  It was really nice when I threw a party since I never had to open up my laptop once, especially since the randomizer worked well between songs.

**Comment by charlieali** [comment]
Would love to see your fork. 

I've been unable to open opc on this build. I get "audioled.opc no such module" even though opc is installed correctly. Any ideas?

**Comment by fht4gn** [comment]
Sparks - as standalone mode ?
In the "Power" mode there is a random spark option - can this be done as a separate mode with the choice of color, quantity and frequency range at which it will work ?

I tried to do it myself, copying the mode and adding a black color mode that would leave only sparks, but this is not the best solution and my knowledge is not enough for more at this time :)

**Comment by MattBrinza** [comment]
Oh boy.  I was working on this from back during the "Christmas" update from the other repo.  Everything has changed since then and its been pretty difficult porting all of my janky changes over.

**Comment by not-matt** [comment]
Hi all! Been a couple weeks, and I think it's time for a quick update. Personal life has soaked up most of my free time lately, hence the inactivity, but this doesn't mean I haven't been thinking about where to steer this project. What I've started working on is beat recognition, building on the existing beat detection framework. What I hope to create is a much more advanced set of tools to pick up patterns in songs' beats (including more irregular beat patterns, say in hip hop), meaning that effects can be much more interesting.

The most basic idea is, for example, your typical four to the floor housy song, where on every bar (rather than beat), something cool happens. Lights flash, whatever, give me some ideas!

Matthew

**Comment by charlieali** [comment]
,Here's the best way I could show an idea for a strobe clap, to go with the beat detection on big claps or beats. There is also a template to share ideas better.

**Comment by charlieali** [comment]
Now that I've been testing this for a bit I have to say it looks great under 256 leds but the more leds you have the harder it is for the effects to fill the strip or matrix.

**Comment by nhmann** [comment]
Hey, I was the guy who had problems with audio-reactive-led-strip on a raspberry pi.
I moved to this project.
When I am able to get it working on a pi, I would write a tutorial for the other users.

Another idea (maybe to much):
it would be nice if you could control the lights with hue for the basic stuff (plain colors).
I think many of us have an existing hue system and it would be nice if the strips would also be controllable for basic lighting.
It could auto change to the audio reactive mode when music starts to play.

**Comment by tarthim** [comment]
@not-matt I've been considering something similair. It would be fantastic if the normal lights react on the instrumental+voice parts on the song, and then these flicker brighter/less bright depending on the beat. However, I'm not sure how one would go about doing that.

Thanks for all your work btw! I set this up yesterday and it's working super well.

**Comment by charlieali** [comment]
Quick thing I found, not a bug just neat. I was trying the dev build on my matrix with a fadecandy but I forgot I was already running Scott Lawson's experimental fork, so when i ran this and then turned the LEDs on I thought I had broken smething. I shut the program down and realized it was still reacting to music. How curious. That's when i realized the other code was running too. I turned yours back on and played with the settings a bit, producing a cool visualization. Since this fadecandy runs via a server and this just sends the data to it, the server can send multiple signals to the LEDs at once. And since the fadecandy itself is meant to dither and switch fast, both visualizations ran simultaneously at 60 fps. So I got curious. I booted up a raspi and loaded this and had it communicate with the fcserver. I now had 3 visualizations going at once with no hiccups. It's really cool. If you set all 3 to spectrum and have each one and set them to different frequencies it looks really cool. Same with running spectrum+wave+scroll all at the same time.  I will try to borrow my works camera this weekend to take some non-phone video of it.


Another thing I was messing around with is in the server config itself, you can set each data pin to be a different color (rgb/brg/grb etc) which obviously changes the color of the strand, so you can have 1 strand be true rbg and others have opposite colors and whatnot. Makes for some nifty fiddling. The Raspi can run everything standalone and the nodemcu has wifi but you should all try a fadecandy, its great.

**Comment by tarthim** [comment]
@charlieali Definitely would like to see a video!

**Comment by Yariv-h** [comment]
An cool feature that can be added - whenever you choose an effect, you have it own customizations on the bottom panel, an automation for the customization or automatic periodic change in the scroll bar can archive a very cool effect, as well auto switch between effect or create a "seen"  that you can set a timer for each effect and select which one will be next end etc..

**Comment by not-matt** [comment]
That's essentially what I'm working on with the beat pattern recognition -
except you're not setting it on a timer (that's no problem to implement),
the lights just seem to change naturally to fit the song.

**Comment by Yariv-h** [comment]
cool, you should consider to "change" the effect itself, lets say add / subtract blue or any other color on the current effect according to the sound, in addition, to change between effects.

Another thing to have in mind, if you will create an API to control the effect, a new modern UI can be created over small http server that will be accessible over the wifi in case you sitting outside and playing music with the leds instead of been on the computer. as well the http ui will have much more flexibility and easy to maintain. i can help with that.

**Comment by blassbasti** [comment]
@nhmann 

> When I am able to get it working on a pi, I would write a tutorial for the other users.
any news on this? I don't really get the thing to work on a standalone raspberry pi

**Comment by Master-P96** [comment]
@not-matt 
Hi, first of all very nice continuation of Scotts project! I've testet it with Anaconda and an ESP, it worked like a charm. 
But actually I want to use it standalone on a (headless) RaspberryPi3.
I was wondering if you could create a Raspberry-Pi version without any GUI-stuff. 
Because after many attemps I was not able to install all the needed packages (especially pyqt5+Python3.6) on Raspbian (jessie/stretch).

**Comment by Mattallmighty** [comment]
@not-matt Created ESP8266 tutorial for /Systematic-LEDs/ README: https://youtu.be/W4jaAgjfvG8

**Comment by Yariv-h** [comment]
@Mattallmighty Thanks!

**Comment by Mattallmighty** [comment]
Hey @not-matt Hope your exams went well 👍 
I have a box RGB ws2812b box as attached. Would be cool to get this supported using my ESP8266 in the future 😃 

![mokungit-16x16-8x32-8x8-pixels-sk6812-ws2812b-digital-flexible-led-panel-individually-addressable-full-dream-color jpg_640x640](https://user-images.githubusercontent.com/20179600/41331010-c3826256-6f2a-11e8-91e6-8d7c536a9ef6.jpg)

**Comment by not-matt** [comment]
Hi all! Got results, all went well (thank goodness) This update brings support for custom colours, audio input selection, and easy board control.

Coming up: 

- Custom gradients for multicolour modes
- Better interface

**Comment by blassbasti** [comment]
hey matt! that's really great to hear!

I already have an idea/request for another feature! it would be great to add multiple devices/strips but control them all from the same UI and not from seperate UIs for every device!

greets

**Comment by not-matt** [comment]
@blassbasti Yeah I've got plans for that. I've got two strips and it's a pain in the ass!

**Comment by not-matt** [comment]
@Mattallmighty Added your tutorial to readme, good stuff! Cheers for that :)

**Comment by not-matt** [comment]
@Master-P96 I'm afraid you need PyQt installed for settings persistence. Headless pi is partially supported already but you can't change the effect or anything (yet)

**Comment by not-matt** [comment]
@Yariv-h Nice idea! That kind of thing will be around when I've made better beat recognition. The idea is to change the effect/setting/whatever on the first beat of every 4 bars or something.

**Comment by not-matt** [comment]
@charlieali Now _that's_ a cool effect. Will be added!

**Comment by f0rmr** [comment]
Is it possible to run this project with ws2801 pixels ?
Want to try it with my 10*10 led matrix.

**Comment by blassbasti** [comment]
yeah true if you have strips with a different length it wouldn't work that way!

in my case i would address 6 strips with 72 pixels each, so it should work, shouldn't it?
I really appreciate your eager working on this project!!

**Comment by not-matt** [comment]
Oh that should be fine, give it a shot and let me know what happens!

**Comment by Moonbase59** [comment]
Hey! I wonder if this project should be steered into a slightly different direction. Don’t understand me wrong, it’s absoluteley awesome (though I cant get it to work under Linux and have to use the previous adaption https://github.com/not-matt/audio-reactive-led-strip), but I have some own ideas and would like to share these and discuss:

1. I always like things to adhere to some standards, so my lights are either controlled via Home Automation (mostly MQTT, some ZigBee), or by the more professional DMX512 (mostly, E1.31 in my case) protocol. Your software, briliant as it is, has some limitations (like 256 LEDs/strip on a microcontroller), and uses its own protocol.

2. I’d _love_ to see the two parts (the great audio controller part vs. the LED strip controlling) to be more separated and, preferably, use a protocol like E1.31.

3. We already have great light controllers that can even do DMX (E1.31) _and_ MQTT so I can use my strips for normal lighting using home automation and simply switch them over to a DMX-based show whenever needed. See https://github.com/forkineye/ESPixelStick for instance.

4. DMX solutions seem to lack a little on the "live audio control" side, so your audio controller would be a perfect fit _if_ it could output E1.31, for example. It could then output to the pixelstick or any other DMX device, and you could concentrate on the real great part, i.e. the audio controller and effects stuff.

5. It would have to be determined if DMX over E1.31 is actually fast enough to do whatever you do today, regarding audio effects. I did some testing with QLC+ and xLights and I think it might be feasible.

6. Separating audio control, protocol, and output devices this way would make Systematic-LEDs even more versatile and also more easily allow for controlling more than one device.

7. And it would make all of our investments in LED strips and other lighting equipment much more awarding, since we could alternatively use the same devices for a) "normal" lighting (smart home), b) DMX-based shows, and c) the kind of audio-reactive party lighting you provide.

What does everybody think? (Well, you asked for it ;-))

**Comment by Moonbase59** [comment]
@blassbasti: I wouldn’t go for general multicasting, especially not within the home network. DMX using E1.31 can do that, and I’ve seen awful network congestion, even on special separate show networks, up to failure of multiple devices during a show.

I even use uncicast in show networks now.

**Comment by ahodges9** [comment]
My entire setup is E1.31 based (ESPixelStick to be precise) and your vison aligns closely with my goals. I use HomeAssistant to control the ESPixelSticks typically (Just recently got in my firmware changes into the official forkineye repro that adds much better MQTT support!) and want to extend upon that to have HomeAssistant to be able to talk to a dedicated RaspberryPi to enable network driven effects (i.e. synced non-reactive effects, more complex effects than ESP01 can handle, audio reactive effects, and video reactive effects).

As for E1.31 I have hard a PR  (#18 E1.31 Support) active here for a while that adds E1.31 support and it works great from my testing (oddly better than pure UDP). E1.31 is 100% capable of support this as is basically just a few extra bytes of header info and a little more processing on the client side. It has a ton of benefits such as packet drop detection and cross device synchronization. The only issues I have found are with DMX512 specifically where the 512 channel limit means some of my strips span universes and thus require two packets giving it more overhead and higher risk of dropping a complete frame. Larger universe sizes is on my list of things to look into though.

As you have noticed, I have already started my own controller "LedFx" that moves towards headless controller ready to be driven by other homeautomation platforms. I just got initial REST and Websocket support that should be fairly easy to integrate with most applications. It is still missing a lot of these effects, but does have a few proof of concept audio reactive ones. If you have a large E1.31 network it would be great to get help testing performance. I only have 4 controllers spanning ~1500 channels which isn't much compared to others

**Comment by Mattallmighty** [comment]
Hey Team,

I have been trying to create scenes/preset at the moment. 
What I am thinking on is hat can press a keyboard shortcut or button highlight below.
![sceneset](https://user-images.githubusercontent.com/20179600/44732279-0af34200-ab39-11e8-8903-d08eeea054ab.PNG)

What will happen is it will load the save file or scene/preset. Keen to know your feedback?

Also another option, is when you add the deice, you can say slave RGB making the transitions between lights seamless. 

Im not a programmer, but learning.

**Comment by blassbasti** [comment]
@Moonbase59 well i use a seperate network just for the visualization. since i have only 7 (1 laptop, 6 ESPs) devices connected to it there shouldn't be any bandwidth issue i think!

**Comment by Yariv-h** [comment]
Would be possible to add support to WS2811 with the upcoming roadmap features?
